crates/flightsql_datafusion/src/lib.rs

```diff
@@ -0,0 +1,465 @@
+        Ok(self)
+    }
+
+    fn execute(
+        &self,
+        _partition: usize,
+        _context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        let sql = match self.sql().map_err(to_execution_error) {
+            Ok(sql) => sql,
+            Err(error) => return Err(error),
+        };
+
+        Ok(Box::pin(StreamConverter::new(
+            self.client.clone(),
+            sql.as_str(),
+            self.schema(),
+        )))
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn to_stream(
+    client: FlightSqlServiceClient<Channel>,
+    sql: &str,
+) -> impl Stream<Item = Result<RecordBatch>> {
+    let client = client.clone();
+    let sql = sql.to_string();
+    stream! {
+    match query(client, sql).await {
+        Ok(stream_opt) => {
+            if let Some(mut stream) = stream_opt {
+                    while let Some(batch) = stream.next().await {
+                        match batch {
+                            Ok(batch) => yield Ok(batch),
+                            Err(error) => yield Err(FlightSQLError::ArrowFlight { source: error })
+                        }
+                    }
+                }
+            },
+            Err(error) => yield Err(error)
+        }
+    }
+}
+
+struct StreamConverter {
+    stream: Pin<Box<dyn Stream<Item = Result<RecordBatch>> + Send>>,
+    schema: SchemaRef,
+}
+
+impl StreamConverter {
+    fn new(client: FlightSqlServiceClient<Channel>, sql: &str, schema: SchemaRef) -> Self {
+        let stream = to_stream(client, sql);
+        Self {
+            stream: Box::pin(stream),
+            schema,
+        }
+    }
+}
+
+impl RecordBatchStream for StreamConverter {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+impl Stream for StreamConverter {
+    type Item = datafusion::common::Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        match self.stream.poll_next_unpin(cx) {
+            Poll::Ready(Some(Ok(record_batch))) => Poll::Ready(Some(Ok(record_batch))),
+            Poll::Ready(Some(Err(error))) => Poll::Ready(Some(Err(to_execution_error(error)))),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn to_execution_error(e: impl Into<Box<dyn std::error::Error>>) -> DataFusionError {
+    DataFusionError::Execution(format!("{}", e.into()).to_string())
+}
+
+async fn query(
+    mut client: FlightSqlServiceClient<Channel>,
+    query: String,
+) -> Result<Option<FlightRecordBatchStream>, FlightSQLError> {
+    let flight_info = client
+        .clone()
+        .execute(query, None)
+        .await
+        .map_err(|e| FlightSQLError::ArrowFlight { source: e.into() })?;
+
+    if let Some(ep) = flight_info.endpoint.first() {
```

Comment:
I don't think we should assume that all Flight SQL servers will just have a single endpoint. We only have a single endpoint, but this code will break if connected to a different Flight SQL server that does have multiple.

Reply:
This is implemented in dataconnector/flightsql.rs